### PR TITLE
v1.1.0

### PR DESCRIPTION
**Pull request by rstoneback**
Includes support for Constellations. Contains bug fixes in orbit reporting, file iteration, and __setitem__.  Includes first support for satellite simulations. Satellite orbit (sgp4 and pyEphem) is simulated using orbit information provided by Two Line Elements (TLEs). Satellite positions is used to drive thermosphere, ionosphere, and geomagnetic models in pyglow. Vector quantities are also projected onto the spacecraft frame (assuming ram pointing). pysat_sgp4 was added to the documentation.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,14 @@
 language: python
 python:
   - "2.7"
-  #- "3.3"
-  - "3.4"
-  - "3.5"
   - "3.6"
```
Comment (aburrell) [inline .travis.yml]: Is support for 3.4 and 3.5 dropped everywhere now?  Or does it break something in particular?

File: README.md [hunk 2]
```diff
@@ -17,6 +17,8 @@ satellite based measurements it aims to support all instruments in space science
 
 Full [Documentation](http://pysat.readthedocs.io/en/latest/index.html)
 
+JGR-Space Physics [Publication](https://doi.org/10.1029/2018JA025297)
+
```
Comment (aburrell) [inline README.md]: :)

File: README.md [hunk 3]
```diff
@@ -48,4 +50,4 @@ pip install pysat
   ```
   pip install netCDF4
   ```
-  
\ No newline at end of file
+
```
Comment (aburrell) [inline README.md]: PEP8: ' -> "

File: pysat/_constellation.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                                ans[zk][yj][0].extend(
+                                    data_considered.ix[yindex, data_label[zk]].tolist())
+
+        # Now for the averaging.
+        # Let's, try .. packing the answers for the 2d function.
+        numx = 1
+        xarr = np.arange(numx)
+        binx = None
+
+        # TODO modify output
```
Comment (aburrell) [inline pysat/_constellation.py]: How does the output need to be modified?

File: pysat/_constellation.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+        # Transform output
+        output = {}
+        for i, label in enumerate(data_label):
+            median = [r[0] for r in out_2d[label]['median']]
+            count  = [r[0] for r in out_2d[label]['count']]
+            dev    = [r[0] for r in out_2d[label]['avg_abs_dev']]
+            output[label] = {'median':  median,
+                             'count':   count,
+                             'avg_abs_dev': dev,
+                             'bin':     out_2d[label]['bin_y']}
```
Comment (aburrell) [inline pysat/_constellation.py]: Odd whitespace use here.

File: pysat/instruments/pysat_testadd2.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    for i in range(len(data['mlt'])):
+        dummy1.append(-i)
+    long_int = (data['longitude']/15.).astype(int)
+    data['dummy1'] = dummy1
+    data['string_dummy'] = ['test']*len(data)
+    data['unicode_dummy'] = [u'test'] * len(data)
+    data['int8_dummy'] = np.array([1] * len(data), dtype=np.int8)
+    data['int16_dummy'] = np.array([1] * len(data), dtype=np.int16)
+    data['int32_dummy'] = np.array([1] * len(data), dtype=np.int32)
+    data['int64_dummy'] = np.array([1] * len(data), dtype=np.int64)
```
Comment (aburrell) [inline pysat/instruments/pysat_testadd2.py]: Would it be better to use np.ones?

File: pysat/instruments/pysat_testadd3.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    data['longitude'] = longitude
+
+    # create latitude area for testing polar orbits
+    latitude = 90.*np.cos(np.mod(uts_root+num_array, 5820)*(2.*np.pi/5820.)) 
+    data['latitude'] = latitude
+    
+    # do slt, 20 second offset from mlt
+    uts_root = np.mod(time_delta.total_seconds()+20, 5820)
+    data['slt'] = np.mod(uts_root+num_array, 5820)*(24./5820.)
+
```
Comment (aburrell) [inline pysat/instruments/pysat_testadd3.py]: This conversion has appeared a few times.  Would it be worth having a little function for it?

File: pysat/instruments/pysat_testsmall2.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    data['dummy2'] = long_int
+    data['dummy3'] = mlt_int + long_int*1000.
+    data['dummy4'] = num_array
+    data['string_dummy'] = ['test']*len(data)
+    data['unicode_dummy'] = [u'test'] * len(data)
+    data['int8_dummy'] = np.array([1] * len(data), dtype=np.int8)
+    data['int16_dummy'] = np.array([1] * len(data), dtype=np.int16)
+    data['int32_dummy'] = np.array([1] * len(data), dtype=np.int32)
+    data['int64_dummy'] = np.array([1] * len(data), dtype=np.int64)
+    # print (data['string_dummy'])
```
Comment (aburrell) [inline pysat/instruments/pysat_testsmall2.py]: Same as before, wrt np.ones

File: .travis.yml (hunk 1 above)
Comment (rstoneback) [reply .travis.yml]: 3.4 is stuck on an older version of pandas. Can start to fix some of the deprecation warnings by getting rid of that. Dropping 3.5 is a matter of convenience. Reduces time on Travis. With 3.7 out may consider testing 2.7 and 3.7.

File: README.md (hunk 2 above)
Comment (rstoneback) [reply README.md]: Agreed :)

File: pysat/instruments/pysat_testadd2.py (hunk 6 above)
Comment (rstoneback) [reply pysat/instruments/pysat_testadd2.py]: Fixed.

**Comment by rstoneback** [comment]
Coverage differs between python 2 and 3 because of pyglow. The system seems to alternate between which it reports. Coverage on 2.7 is up. Parity will return when pyglow is python 3 compatible.

**Comment by rstoneback** [comment]
Thanks Angeline!

Scrutinizer had like 16 hours, I think there is a temporary coupling problem.

Anyway, the merge is in.

Cheers.

On Jul 6, 2018, at 3:25 PM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


You're good to merge as soon as scrutinizer finishes.

**Pull request merged**